pandas/io/sas/sas7bdat.py

```diff
@@ -0,0 +1,801 @@
+            return True
+        elif len(self._cached_page) != self._page_length:
+            raise Exception("failed to read complete page from file (read %d of %d bytes)",
+                            len(self._cached_page), self._page_length)
+
+        self._read_page_header()
+        if self._current_page_type == _page_meta_type:
+            self._process_page_metadata()
+        if self._current_page_type not in [_page_meta_type, _page_data_type] + _page_mix_types:
+            return self._read_next_page()
+
+        return False
+
+    def _decompress(self, row_length, page):
+        page = np.frombuffer(page, dtype=np.uint8)
+        if self.compression == _rle_compression:
+            return _rle_decompress(row_length, page)
+        elif self.compression == _rdc_compression:
+            return _rdc_decompress(row_length, page)
+        else:
+            raise ValueError("unknown SAS compression method: %s" % self.compression)
+
+    def _process_byte_array_with_data(self, offset, length):
+
+        if (self.compression != "") and (length < self.row_length):
+            source = self._decompress(self.row_length, self._cached_page[offset:offset + length])
+        else:
+            source = self._cached_page[offset:offset + length]
+
+        s = 8 * self._current_row_in_chunk_index
+        for j in range(self.column_count):
+            length = self._column_data_lengths[j]
+            if length == 0:
+                break
+            start = self._column_data_offsets[j]
+            end = start + length
+            temp = source[start:end]
+            if self.columns[j].ctype == 'd':
+                m = 8 - length
+                if self.byte_order == "<":
+                    self._chunk[j][s+m:s+8] = np.frombuffer(temp, dtype=np.uint8)
+                else:
+                    self._chunk[j][s:s+length] = np.frombuffer(temp, dtype=np.uint8)
+            elif self.columns[j].ctype == 's':
+                self._chunk[j][self._current_row_in_chunk_index] = temp.rstrip(b"\x00 ").decode()
+            else:
+                raise ValueError("unknown column type: %s" % self.columns[j].ctype)
+
+        self._current_row_on_page_index += 1
+        self._current_row_in_chunk_index += 1
+        self._current_row_in_file_index += 1
+
+    def _chunk_to_dataframe(self):
+
+        n = self._current_row_in_chunk_index
+        m = self._current_row_in_file_index
+        ix = range(m - n, m)
+        rslt = pd.DataFrame(index=ix)
+
+        for j in range(self.column_count):
+
+            name = self.column_names[j]
+
+            if self.column_types[j] == 'd':
+                rslt[name] = self._chunk[j].view(dtype=self.byte_order + 'd')
+                rslt[name] = np.asarray(rslt[name], dtype=np.float64)
+                if self.convert_dates and (self.column_formats[j] == "MMDDYY"):
+                    epoch = pd.datetime(1960, 1, 1)
+                    rslt[name] = epoch + pd.to_timedelta(rslt[name], unit='d')
```

Comment:
it is pretty inefficient to do this. much better is something like this.

```
l = []
for j in range(self.column_count):
    # create an array of the appropriate dtype
    arr = ....
    l.append(arr)
result = pd.concat(l, keys=self.column_names)
```
will have at most 1 copy of everything.

the current method could potentially have a fair number of copies (this is in the current impl of pandas
eventually this will not be the case, but that's in the future :)